### PR TITLE
feat: API usage mapping proof of concept

### PR DESCRIPTION
**Pull request by marqode**
## Summary

This PR introduces a non-intrusive API Usage Mapper and Contract Coverage Proof-of-Concept. By hooking into MSW at the network layer during Vitest execution, it transparently records, normalizes, and profiles front-end API interactions without modifying application code or existing test assertions. The output provides a token-optimized, inferred API map designed for automated coverage validation and AI agent ingestion.

### Operational Quickstart
Run the test suite normally to generate the data stream:
`pnpm coverage`

Compile the aggregated JSON report:
`pnpm coverage-report`

Or do both with the same command:
`pnpm coverage:full`

- Telemetry Log Stream: src/tests/msw-api-registry.jsonl
- Aggregated Contract Map: src/tests/msw-contract-coverage.json

### Key Changes
- Pipeline Instrumentation: Configured vitest.config.ts to implement a master process globalSetup script (wipes old logs) and worker thread setupFiles (attaches MSW lifecycle listeners).

- Thread-Safe Telemetry Engine: Streams intercepted traffic atomically to disk to prevent write collisions across parallel Vitest workers. Uses .clone() on network streams to ensure zero interference with running components or Axios/Fetch instances.

- Heuristic Aggregator Script: Cleans dynamic path IDs, hex values, and slugs (e.g., /v1beta1/operations/mirror-abc collapses to /v1beta1/operations/:id) to prevent path fragmentation.

- Payload Deduplication: Uses SHA-1 structural hashing on write payloads (POST, PUT, PATCH). If a mutation runs repeatedly with the same shape, only one contract example is preserved per HTTP status code to minimize context window bloat.

### Downstream AI & Coverage Strategy
Because our target repositories currently have low OpenAPI documentation coverage (~5-10%), this client-side mapping serves as a baseline profile that allows an automated agent or pipeline to:

- Identify Gaps: Instantly cross-reference backend routers against the map to find untested route patterns (totalHits: 0).

- Expose Blindspots: Detect routes that only verify happy-path responses (200 OK) while completely missing failure assertions (4xx or 5xx validation states).

## Release Impact

Pick the target branch (see [RELEASES.md](../RELEASES.md) for the full model):

- [ ] `dev` — internal testing → publishes to `ppa-build-dev`
- [X] `main` — next beta → publishes to `ppa-build`
- [ ] `release/YY.MM` — point release on a maintained cycle → publishes to `ppa-build-YY.MM` (and to `ppa-build-stable` if this is the currently-promoted branch). Specify cycle: `release/____`

Change type (tick one):

- [ ] Patch (fix)
- [X] Minor (feature)
- [ ] Major (breaking)

> The change-type label is informational and only affects how the entry is rendered in the CHANGELOG. The actual version is computed from the branch and CalVer cycle — `pnpm changeset`'s `patch`/`minor` choice does not influence it.

## Checklist

- [X] **Changeset added** — I have run `pnpm changeset` (or `pnpm changeset --empty` if no CHANGELOG line is warranted) and committed the resulting `.md` file. Required for PRs targeting `main` and `release/*`; enforced by the `Changeset check` workflow.
- [ ] **UI verified** — I have verified the changes locally.
- [X] **Linting clean** — No linting errors are present (especially in `scripts/`).

## Versioning Reminder

> [!IMPORTANT]
> Landscape UI uses **CalVer** (`YY.MM.Point.Build`). Version derivation by branch:
>
> - `main` / `point/*` → `YY.MM.0.<run>-beta`
> - `dev` → `YY.MM.0.<run>-dev`
> - `release/YY.MM` → `YY.MM.1.<run>` (cycle pinned by branch name)

### CODE REVIEW

**Comment by yurii-vasyliev** [comment]
Follow-up POC on top of this branch: #708 — replaces the heuristic `normalizeRoute` with pluggable contract connectors (Go OpenAPI spec, v1 RPC actions, MSW handler patterns), labels every route with its backend generation (v1/v2/go), and adds a hand-maintained migration map. Full-suite run resolves all 10,482 recorded interactions to declared contracts with zero heuristic fallbacks. Details and rationale in the PR description.

File: src/tests/setup.ts [hunk 1]
```diff
@@ -13,12 +15,82 @@ import server from "./server";
 
 expect.extend(matchers);
 
-// Async DOM waits (findBy*, waitFor, waitForElementToBeRemoved) default to
-// 1000ms, which is too tight under the full coverage run on CI — especially on
-// the merge-queue branch where workers contend for CPU. Give them more headroom
-// so load-sensitive tests don't flake (and silently drop the PR from the queue).
+// Configure timeout limits for CI headroom
```
Comment (yurii-vasyliev) [inline src/tests/setup.ts]: This rewrite drops three comment blocks that exist on `main` and are load-bearing:

1. The original comment here explains *why* `asyncUtilTimeout` is 5000 (CI merge-queue CPU contention causing flakes that silently drop PRs from the queue) — "Configure timeout limits for CI headroom" loses the rationale.
2. The `scrollIntoView` polyfill note (jsdom doesn't implement it; needed for keyboard-nav components).
3. The `Blob.prototype.stream` polyfill explanation — that one documents a subtle MSW/undici/axios flake chain that took real debugging to find; without it the polyfill looks removable.

Since `main` hasn't touched those lines, git will merge the deletions silently. Could you restore the original comments?

File: .changeset/ripe-cars-fall.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+---
+"landscape-ui": minor
```
Comment (yurii-vasyliev) [inline .changeset/ripe-cars-fall.md]: The changeset declares a `minor` bump but has no description text, so the generated CHANGELOG entry would be blank. Either add a one-liner (e.g. "Add MSW-based API contract coverage tooling for tests") or — since this is test-only tooling with no shipped-code impact — consider `pnpm changeset --empty` instead, which skips the CHANGELOG entirely.

**Review by yurii-vasyliev — COMMENTED** [review]
Reviewed the POC with the connector follow-up (#708) now merged in. The approach is solid and the full suite is green locally — but a few blockers before this is approvable:

1. **Deleted `setup.ts` comments** (inline comment below) — three explanatory blocks from `main` are dropped in this diff and would merge away silently.
2. **Empty changeset** (inline comment below) — the `minor` bump has no description, so the CHANGELOG entry would render blank.
3. **Stale PR description** — it still documents the original heuristic normalizer (`dynamicParents`, SHA-1 dedup, `npx tsx` quickstart). Since #708 replaced that with contract connectors (OpenAPI / v1 actions / MSW manifest, plus the migration map), could you refresh the description? `pnpm run coverage-report` / `coverage:full` are the entry points now.

FYI: the Prettier CI failure was from the #708 files — fixed and pushed directly to this branch as 96794cbc8.

File: src/tests/setup.ts (hunk 1 above)
Comment (marqode) [reply src/tests/setup.ts]: good call, done

File: src/tests/setup.ts [hunk 3]
```diff
@@ -21,6 +23,90 @@ expect.extend(matchers);
...
+      responsePayload,
+    };
+
+    // Concurrent appends from parallel Vitest workers may theoretically interleave,
+    // but JSONL line-length is short enough that OS-level write buffering makes
+    // corruption extremely unlikely in practice.
+    await fs.promises.appendFile(
+      REGISTRY_PATH,
+      JSON.stringify(logEntry) + "\n",
+    );
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/setup.ts]: Per-worker output files would be overkill for this feature. Added a `truncateForLog` helper to enforce write atomicity so written bytes are never interleaved.

File: src/tests/contract-coverage/paths.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import path from "path";
+
+const TESTS_DIR = path.resolve(import.meta.dirname, "..");
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/contract-coverage/paths.ts:3]: consistent with other usage in the codebase

File: src/tests/setup.ts [hunk 5]
```diff
@@ -21,6 +23,94 @@ expect.extend(matchers);
...
+      status: response.status,
+      requestPayload,
+      responsePayload,
+    };
+
+    // Concurrent appends from parallel Vitest workers may theoretically interleave,
+    // but JSONL line-length is short enough that OS-level write buffering makes
+    // corruption extremely unlikely in practice.
+    await fs.promises.appendFile(
+      REGISTRY_PATH,
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/setup.ts]: resolution noted on below comment

File: src/tests/setup.ts [hunk 6]
```diff
@@ -21,6 +23,103 @@ expect.extend(matchers);
...
+// effectively atomic across concurrent Vitest worker threads. Payloads larger
+// than this are replaced with a sentinel so the line stays well under 4 KB.
+const MAX_PAYLOAD_BYTES = 2048;
+
+function truncateForLog(payload: unknown): unknown {
+  if (payload === null) return null;
+  const serialized = JSON.stringify(payload);
+  if (serialized.length <= MAX_PAYLOAD_BYTES) return payload;
+  return { __truncated: true, originalBytes: serialized.length };
+}
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/setup.ts:57]: - JSON.parse is already run on payload so `.stringify` will never throw
- added `Buffer.byteLength(serialized, "utf8");` for extra safety factor

File: package.json [hunk 7]
```diff
@@ -68,10 +70,12 @@
...
     "stylelint-config-standard-scss": "17.0.0",
     "stylelint-order": "8.1.1",
+    "tsx": "^4.23.1",
     "typed-css-modules": "0.9.1",
     "typescript": "6.0.3",
     "typescript-eslint": "8.60.1",
     "vite": "8.0.16",
-    "vitest": "4.1.8"
+    "vitest": "4.1.8",
+    "yaml": "^2.9.0"
```
[the remark this answers is not in the document]
Comment (marqode) [reply package.json]: out of scope

File: package.json [hunk 8]
```diff
@@ -68,10 +70,12 @@
     "stylelint": "17.12.0",
     "stylelint-config-standard-scss": "17.0.0",
     "stylelint-order": "8.1.1",
+    "tsx": "^4.23.1",
```
[the remark this answers is not in the document]
Comment (marqode) [reply package.json]: out of scope

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (marqode) [reply package.json]: out of scope

File: src/tests/setup.ts [hunk 9]
```diff
@@ -21,6 +23,104 @@ expect.extend(matchers);
...
+// than this are replaced with a sentinel so the line stays well under 4 KB.
+const MAX_PAYLOAD_BYTES = 2048;
+
+function truncateForLog(payload: unknown): unknown {
+  if (payload === null) return null;
+  const serialized = JSON.stringify(payload);
+  if (Buffer.byteLength(serialized, "utf8") <= MAX_PAYLOAD_BYTES)
+    return payload;
+  return { __truncated: true, originalBytes: serialized.length };
+}
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/setup.ts:57]: in practice this would never happen for the ASCII characters handled here

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (marqode) [reply package.json]: out of scope

[the remark this answers is not in the document]
Comment (yurii-vasyliev) [reply package.json]: These two lines (`tsx` here and `yaml` further down) are added by this PR, and every other `devDependency` in the file is pinned exactly. It's worth applying the proposed changes.

File: src/tests/setup.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (yurii-vasyliev) [reply src/tests/setup.ts:57]: It's still a tiny inconsistency, I'd apply it.

**Review by yurii-vasyliev — DISMISSED** [review]
LGTM, just a few tiny things in the comments.

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (marqode) [reply package.json]: Thanks! I just assumed we always had tsx here, should've double checked.

File: src/tests/setup.ts [hunk 10]
```diff
@@ -101,7 +201,8 @@ beforeAll(() => {
   resetScreenSize();
 });
 
-afterAll(() => {
+afterAll(async () => {
+  await Promise.allSettled(pendingLogs);
   server.close();
   restoreRangeBoundingClientRect();
 });
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/setup.ts:208]: fixed by moving `server.close()` to the top

File: src/tests/contract-coverage/aggregate.ts [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    layers: Partial<Record<MigrationLayer, MigrationLayerStatus>>;
+  }[];
+  warnings: string[];
+}
+
+function hashPayload(payload: unknown): string {
+  if (payload === null || payload === undefined) return "empty";
+  const str = typeof payload === "string" ? payload : JSON.stringify(payload);
+  return crypto.createHash("sha1").update(str).digest("hex");
+}
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/contract-coverage/aggregate.ts:65]: not worth the added complexity, shouldn't be an issue in practice

File: src/tests/setup.ts [hunk 12]
```diff
@@ -21,6 +23,104 @@ expect.extend(matchers);
...
+  });
+  try {
+    fs.writeFileSync(MANIFEST_PATH, JSON.stringify(manifest, null, 2), {
+      flag: "wx",
+    });
+  } catch (error) {
+    if ((error as NodeJS.ErrnoException).code !== "EEXIST") {
+      throw error;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/setup.ts:115]: handled in global-setup.ts

File: src/tests/setup.ts [hunk 13]
```diff
@@ -21,6 +23,104 @@ expect.extend(matchers);
...
+    };
+
+    await fs.promises.appendFile(
+      REGISTRY_PATH,
+      JSON.stringify(logEntry) + "\n",
+    );
+  } catch (error) {
+    console.error("Failed to write MSW interaction registry entry:", error);
+  }
+}
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/setup.ts:80]: already addressed sufficiently

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (yurii-vasyliev) [reply package.json]: `tsx` was there, but without the caret :)

File: src/tests/global-setup.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import fs from "fs";
+import { MANIFEST_PATH, REGISTRY_PATH } from "./contract-coverage/paths";
+
+export default function setup() {
+  for (const file of [REGISTRY_PATH, MANIFEST_PATH]) {
+    if (fs.existsSync(file)) {
+      fs.unlinkSync(file);
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (marqode) [reply src/tests/global-setup.ts:10]: This feels like the wrong place to clear the report- global-setup only handles the vitest run, while the report is produced by `aggregate-coverage`. If we want to prevent stale reports fgrom being aggregated we could add a check to that script later on.

**Review by yurii-vasyliev — APPROVED** [review]
LGTM

**Review by RyanLoi98 — APPROVED** [review]
LGTM! 

Good work Marc!
